File: vector/src/vector/World.java
package vector;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class World {
	public static final String TEAPOT_FILE_DIRECTORY = "/Users/DaekunKim/Documents/Programming Related/Eclipse Projects/VectorWorld/vector/assets/teapot";
	public static final String TEACUP_FILE_DIRECTORY = "/Users/DaekunKim/Documents/Programming Related/Eclipse Projects/VectorWorld/vector/assets/teacup";
	
	private Vector3[] teapotVertices;
	private int[][] teapotPatches;
	
	private int numTeapotVertices;
	private int numTeapotPatches;
	
	private double rotation;
	private Vector3 rotationAxis;
	private double scale;
	private Vector3 translation;
	
	public World() {
		try {
			InputStream in = new FileInputStream(new File(TEAPOT_FILE_DIRECTORY));
			BufferedReader reader = new BufferedReader(new InputStreamReader(in));
			
			String s;
			
			s = reader.readLine(); // Number of patches;
			numTeapotPatches = Integer.parseInt(s);
			teapotPatches = new int[numTeapotPatches][];
			
			for (int i = 0; i < numTeapotPatches; i++) {
				s = reader.readLine();
				
				String[] patchIndexStrings = s.split(",");
				
				teapotPatches[i] = new int[patchIndexStrings.length];
				for (int j = 0; j < patchIndexStrings.length; j++) {
					teapotPatches[i][j] = Integer.parseInt(patchIndexStrings[j]);
				}
			}
			
			s = reader.readLine(); // Number of vertices
			numTeapotVertices = Integer.parseInt(s);
			teapotVertices = new Vector3[numTeapotVertices];
			
			for (int i = 0; i < numTeapotVertices; i++) {
				s = reader.readLine();
				
				String[] vertexIndexStrings = s.split(",");
				teapotVertices[i] = new Vector3(Double.parseDouble(vertexIndexStrings[0]),
											Double.parseDouble(vertexIndexStrings[1]),
											Double.parseDouble(vertexIndexStrings[2]));
			}
			
			reader.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
		
		rotation = 0;
		rotationAxis = new Vector3(1, 1, 1);
		scale = 50;
		translation = new Vector3(200, 200, 0);
	}
	
	/**
	 * Updates the world and its objects, and renders it on renderer
	 * @param renderer Renderer object for the objects in the world to be rendered to
	 */
	public void update(Renderer renderer) {
		Vector3[] manipulatedVertices = new Vector3[teapotVertices.length];
		
		for (int i = 0; i < manipulatedVertices.length; i++) {
			manipulatedVertices[i] = teapotVertices[i].scalarMultiply(scale);
			manipulatedVertices[i].rotate(rotationAxis, rotation);
			manipulatedVertices[i] = manipulatedVertices[i].add(translation);
		}
		
		Mesh teapotMesh = new Mesh(manipulatedVertices, teapotPatches);
		
		renderer.drawMesh(teapotMesh);
		rotation += 90.0 / Display.FPS;
	}
}

File: VectorWorld/world.cpp
#include "main.h"

World::World() {
    ifstream teapotFile;
    teapotFile.open("/Users/DaekunKim/Documents/Programming Related/VectorWorld/VectorWorld/assets/teapot");
    
    string s;
    getline(teapotFile, s);
    numTeapotPatches = atoi(s.c_str());
    teapotPatches = new int*[numTeapotPatches];
    
    for (int i = 0; i < numTeapotPatches; i++) {
        teapotPatches[i] = new int[NUM_VERTICES_PER_PATCH];
        
        getline(teapotFile, s);
        stringstream ss(s);
        
        string token;
        for (int j = 0; getline(ss, token, ','); j++) {
            // Putting getline in the incremental side will do no good, since it will only get executed when the code block below is done executing, which means that the initial value of 'token' would be uninitialized, so atoi() will return 0 or other jiberish.
            teapotPatches[i][j] = atoi(token.c_str());
        }
    }
    
    getline(teapotFile, s);
    numTeapotVertices = atoi(s.c_str());
    teapotVertices = new Vector3[numTeapotVertices];
    
    for (int i = 0; i < numTeapotVertices; i++) {
        getline(teapotFile, s);
        stringstream ss(s);
        
        string token;
        
        getline(ss, token, ',');
        teapotVertices[i].x = atof(token.c_str());
        getline(ss, token, ',');
        teapotVertices[i].y = atof(token.c_str());
        getline(ss, token, ',');
        teapotVertices[i].z = atof(token.c_str());
    }
    
    teapotMesh = new Mesh(numTeapotVertices, teapotVertices, numTeapotPatches, teapotPatches);
    
    for (int i = 0; i < numTeapotPatches; i++) {
        for (int j = 0; j < NUM_VERTICES_PER_PATCH; j++) {
            cout << teapotPatches[i][j] << " ";
        }
        cout << endl;
    }
    
    rotation = 0;
    rotationAxis = Vector3(1, 1, 1);
    scale = 50;
    translation = Vector3(SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2, 0);
}

World::~World() {
}

void World::update(Renderer *renderer) {
    Vector3 *manipulatedVertices = new Vector3[teapotMesh->getNumVertices()];
    
    for (int i = 0; i < teapotMesh->getNumVertices(); i++) {
        manipulatedVertices[i] = teapotMesh->vertices[i];
        manipulatedVertices[i] = manipulatedVertices[i] * scale;
        manipulatedVertices[i].rotate(Vector3(1, 0, 0), 270);
        manipulatedVertices[i].rotate(rotationAxis, rotation);
        manipulatedVertices[i] = manipulatedVertices[i] + translation;
    }
    
    int **manipulatedPatches = new int*[teapotMesh->getNumPatches()];
    
    for (int i = 0; i < teapotMesh->getNumPatches(); i++) {
        manipulatedPatches[i] = new int[NUM_VERTICES_PER_PATCH];
        for (int j = 0; j < NUM_VERTICES_PER_PATCH; j++) {
            manipulatedPatches[i][j] = teapotPatches[i][j];
        }
    }
    
    Mesh *resultMesh = new Mesh(teapotMesh->getNumVertices(), manipulatedVertices, teapotMesh->getNumPatches(), manipulatedPatches);
    
    renderer->renderMesh(resultMesh);
    rotation += 90.0 * Time::deltaTime;

//    scale += 0.1;
    
    manipulatedVertices = NULL;
    manipulatedPatches = NULL;
    
    delete resultMesh;
    
//    Vector3 vertices[4];
//    
//    vertices[0] = Vector3(100, 100, 0);
//    vertices[1] = Vector3(200, 100, 0);
//    vertices[2] = Vector3(200, 200, 0);
//    vertices[3] = Vector3(100, 200, 0);
//    
//    renderer->renderPolygon(vertices, 4, 100, 200);
}